HackerRank > Practice > Algorithms > Cut the sticks - solved:

rododin opened the pull request:
- A pretty trivial recursive solution has been implemented.
 - The solution is not effective though. The complexity is `O(2n(n-1))` in worst case, and it leads to extremely huge execution time on huge input arrays.
   Also, the recursion introduces additional constraints disallowing to handle huge input arrays because of the `StackOverflowError`.
   However it still passed all the HackerRank tests and was accepted.
 - Appropriate unit tests have been added.

rododin committed:
HackerRank > Practice > Algorithms > Cut the sticks - solved:
 - A pretty trivial recursive solution has been implemented.
 - The solution is not effective though. The complexity is `O(2n(n-1))` in worst case, and it leads to extremely huge execution time on huge input arrays.
   Also, the recursion introduces additional constraints disallowing to handle huge input arrays because of the `StackOverflowError`.
   However it still passed all the HackerRank tests and was accepted.
 - Appropriate unit tests have been added.

rododin-chesscom reviewed (changes requested):
Let's keep this algorithm, but look for something more effective. E.g. we could probably use the standard Java `sort` algorithm to sort the array, then iterate it for only one time more, removing the duplicates (which follow one-by-one because it's already sorted) and counting the remained elements.

rododin committed:
HackerRank > Practice > Algorithms > Cut the sticks - solved:
 - Another more effective solution has been implemented.
   The complexity is `O(n + O(standardJavaIntegerListSortAlgorithm))`.
   It doesn't introduce any additional constraints compared with the previously implemented recursive `cutTheSticksRecursive(List, List)` algorithm.
 - Unit tests have been added. Now the new algorithm well and quickly passes the test with 1 million elements.

rododin-chesscom commented on modules/hackerrank/src/main/java/org/rododin/algorithms/hackerrank/problem_solving/easy/CutTheSticks.java:
```diff
@@ -0,0 +1,88 @@
+			}
+		}
+		return result;
+	}
```
👍

rododin-chesscom reviewed (approved):
Much better! 👍
